### DISCUSSION TITLE
Can I profile jitted CPU code?

### DISCUSSION THREAD

**Discussion by mvsoom** [jax-ml/jax#19888] [Q&A]
Profiling jitted CPU code with the jax profiler shows a single call to the compiled code. What other tools are there to accomplish this? (Related to jax issue #774.)

**Comment by jakevdp** [comment]
There is some information on profiling here: https://jax.readthedocs.io/en/latest/profiling.html

**Reply by tgbrooks** [reply]
I have the same question as mvsoom and have read and followed the profiling guide (i.e., I'm using `jax.profiler.start_profile()` and then running my code while collecting a trace with TensorBoard). I still only get a single entry per jitted function with no way of knowing how the time is spent within that function. For example, I see on TensorBoard a slew of "PjitFunction(_objective)" function calls but without any indication of where within my `_objective` function the time is spent. I tried adding `@jax.profiler.annotate_function` to my sub functions without any effect.

I am using the CPU only so I haven't used `nvprof` as suggested in #774 . I understand that profiling optimized compiled code can be tricky in any case but I was hoping to have a little more granularity than I'm getting. Perhaps the instructions in the documentation work better when using a GPU?

**Reply by tgbrooks** [reply]
I tried again on a PC with a GPU and found that profiling GPU code provides much richer information than when profiling CPU code. Is that known/intentional? It doesn't seem to be documented.

**Reply by acertain** [reply]
Is it possible to get perf's jit profiling working with jax? a la https://github.com/tensorflow/tensorflow/commit/a45a58b5d8e5db79c8d47b943c3ea73cb4860022

**Reply by joaospinto** [reply]
I'm also interested in profiling CPU workflows. Is the current recommended way of doing so to manually add custom tracing events to my code? Some general support for sampling-based profiling (instead of just tracing) would be super useful. CC @jakevdp

**Comment by cevheck** [comment]
Hi all,

Any updates here? I am trying to optimize a jax.lax.scan function. I see a big discrepancy between the time needed for the first call to the scan and the second one (probably due to caching). I would like to optimize and therefore profile this second (or later) call.

I'm also working on CPU.

**Reply by jakevdp** [reply]
For doing micro-benchmarks like this, you should keep in mind the effects of jit-compilation, asynchronous dispatch, and data placement: https://docs.jax.dev/en/latest/faq.html#benchmarking-jax-code. Some combination of those is probably leading to the slow first call.

**Comment by Gattocrucco** [comment]
I'm having the same problem and I ended up resorting mostly to splitting up my jitted function into smaller jitted functions. This is quite annoying and has side effects, but I couldn't do any better.

Another partial solution I also use is to set the host tracing level to 2, look at auxiliary threads, and use my intuition to match the operations I see in the trace to what my code is doing:

```python
from jax.profiler import ProfileOptions, trace

opts = ProfileOptions()
opts.host_tracer_level = 2

with trace('./jaxtraces', profiler_options=opts):
    ...
```

Then in the trace I find the jit function I want to analyse, and the corresponding operations happening on an auxiliary thread:

<img width="755" height="774" alt="image" src="https://github.com/user-attachments/assets/851367ae-89cd-4614-aeab-ed6c08a615b2" />


If I zoom temporally, I can see various operations running in the auxiliary thread, and knowing my code I can recognize where the most important pieces are coming from:

<img width="1055" height="174" alt="image" src="https://github.com/user-attachments/assets/b3b412e0-2407-466b-b4d0-c9214da08144" />
